### PR TITLE
CMR-9740: Modify kms-* caches refresh logic

### PR DESCRIPTION
**Pull request by jmaeng72**
# Overview

### What is the feature/fix?

To prevent kms-* caches from refreshing if there is a redis connection error or redis related exception

### What is the Solution?

Allow the exception to bubble up to kms calls and then throw the exception or return nil as result

### What areas of the application does this impact?

Any cache using redis (since we are now throwing the exception found), search-app

# Checklist

- [x] I have updated/added unit and integration tests that prove my fix is effective or that my feature works
- [x] New and existing unit and int tests pass locally and remotely with my changes
- [x] I have removed unnecessary/dead code and imports in files I have changed
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings

### CODE REVIEW

File: common-app-lib/src/cmr/common_app/services/kms_fetcher.clj [hunk 1]
```diff
@@ -60,36 +62,50 @@
...
+                ;; if the keyword-scheme-value is nil that means we could not get the KMS keywords
+                ;; in this case use the cached value instead so that we don't wipe out the cache.
+                (if-let [keyword-scheme-value (kms/get-keywords-for-keyword-scheme context keyword-scheme)]
+                  [keyword-scheme keyword-scheme-value]
+                  [keyword-scheme (get kms-cache-value keyword-scheme)])))))
+    (catch Exception e
+      (if (clojure.string/includes? (ex-message e) "Carmine connection error")
+        (do
+          (error "fetch-gcmd-keywords-map found redis carmine exception. Will return nil result." e)
+          nil)
```
Comment (eereiter) [inline common-app-lib/src/cmr/common_app/services/kms_fetcher.clj:83]: Remove the (do    and the nil)
(error returns nil.

Please do this in the kms_lookup namespace as well.

File: common-app-lib/src/cmr/common_app/services/kms_fetcher.clj [hunk 2]
```diff
@@ -60,36 +62,50 @@
...
+          (let [[t1 kms-index] (util/time-execution (fetch-gcmd-keywords-map context))
+                [t2 _] (util/time-execution (cache/set-value cache kms-cache-key kms-index))]
+            (rl-util/log-data-gathering-stats "get-kms-index" kms-cache-key t1)
+            (rl-util/log-redis-write-complete "get-kms-index" kms-cache-key t2)
+            kms-index))))
+    (catch Exception e
+      (if (clojure.string/includes? (ex-message e) "Carmine connection error")
+        (do
+          (error "get-kms-index found redis carmine exception. Will return nil result." e)
+          nil)
```
Comment (eereiter) [inline common-app-lib/src/cmr/common_app/services/kms_fetcher.clj:107]: Remove the (do and the nil)
(error returns nil.

Please do this in the kms_lookup namespace as well.

File: common-app-lib/src/cmr/common_app/services/kms_lookup.clj [hunk 3]
```diff
@@ -240,32 +242,46 @@
...
+                       (let [_ (load-cache-if-necessary context short-name-cache kms-short-name-cache-key)
+                             [tm kwords] (util/time-execution (hash-cache/get-value short-name-cache kms-short-name-cache-key keyword-scheme))]
+                         (rl-util/log-redis-read-complete "lookup-by-short-name" kms-short-name-cache-key tm)
+                         kwords))]
+        (get keywords (util/safe-lowercase short-name))))
+    (catch Exception e
+      (if (clojure.string/includes? (ex-message e) "Carmine connection error")
+        (do
+          (error "lookup-by-short-name found redis carmine exception. Will return nil result." e)
+          nil)
```
Comment (eereiter) [inline common-app-lib/src/cmr/common_app/services/kms_lookup.clj:261]: Remove the (do and the nil)
(error returns nil.

Please do this in the kms_lookup namespace as well.

File: common-app-lib/src/cmr/common_app/services/kms_lookup.clj [hunk 4]
```diff
@@ -240,32 +242,46 @@
...
+          keywords
+          (let [_ (load-cache-if-necessary context location-cache kms-location-cache-key)
+                [tm kwords] (util/time-execution (hash-cache/get-value location-cache kms-location-cache-key (string/upper-case location-string)))]
+            (rl-util/log-redis-read-complete "lookup-by-location-string" kms-location-cache-key tm)
+            kwords))))
+    (catch Exception e
+      (if (clojure.string/includes? (ex-message e) "Carmine connection error")
+        (do
+          (error "lookup-by-location-string found redis carmine exception. Will return nil result." e)
+          nil)
```
Comment (eereiter) [inline common-app-lib/src/cmr/common_app/services/kms_lookup.clj:283]: Remove the (do and the nil)
(error returns nil.

Please do this in the kms_lookup namespace as well.

File: common-app-lib/src/cmr/common_app/services/kms_lookup.clj [hunk 5]
```diff
@@ -280,95 +296,122 @@
...
+                  (let [_ (load-cache-if-necessary context umm-c-cache kms-umm-c-cache-key)
+                        [tm vlue] (util/time-execution (hash-cache/get-value umm-c-cache kms-umm-c-cache-key keyword-scheme))]
+                    (rl-util/log-redis-read-complete "lookup-by-umm-c-keyword-data-format" kms-umm-c-cache-key tm)
+                    vlue))]
+      (get-in value [comparison-map]))
+    (catch Exception e
+      (if (clojure.string/includes? (ex-message e) "Carmine connection error")
+        (do
+          (error "lookup-by-umm-c-keyword-data-format found redis carmine exception. Will return nil result." e)
+          nil)
```
Comment (eereiter) [inline common-app-lib/src/cmr/common_app/services/kms_lookup.clj:320]: Remove the (do and the nil)
(error returns nil.

Please do this in the kms_lookup namespace as well.

File: common-app-lib/src/cmr/common_app/services/kms_lookup.clj [hunk 6]
```diff
@@ -280,95 +296,122 @@
...
+        (get-in value [comparison-map])
+        ;; Check just shortname
+        (-> value
+            (remove-long-name-from-kms-index)
+            (get comparison-map))))
+    (catch Exception e
+      (if (clojure.string/includes? (ex-message e) "Carmine connection error")
+        (do
+          (error "lookup-by-umm-c-keyword-platforms found redis carmine exception. Will return nil result." e)
+          nil)
```
Comment (eereiter) [inline common-app-lib/src/cmr/common_app/services/kms_lookup.clj:353]: Remove the (do and the nil)
(error returns nil.

Please do this in the kms_lookup namespace as well.

File: common-app-lib/src/cmr/common_app/services/kms_lookup.clj [hunk 7]
```diff
@@ -280,95 +296,122 @@
...
+                      (let [_ (load-cache-if-necessary context umm-c-cache kms-umm-c-cache-key)
+                            [tm vlue] (util/time-execution (hash-cache/get-value umm-c-cache kms-umm-c-cache-key keyword-scheme))]
+                        (rl-util/log-redis-read-complete "lookup-by-umm-c-keyword" kms-umm-c-cache-key tm)
+                        vlue))]
+          (get-in value [comparison-map]))))
+    (catch Exception e
+      (if (clojure.string/includes? (ex-message e) "Carmine connection error")
+        (do
+          (error "lookup-by-umm-c-keyword found redis carmine exception. Will return nil result." e)
+          nil)
```
Comment (eereiter) [inline common-app-lib/src/cmr/common_app/services/kms_lookup.clj:383]: Remove the (do and the nil)
(error returns nil.

Please do this in the kms_lookup namespace as well.

File: common-app-lib/src/cmr/common_app/services/kms_lookup.clj [hunk 8]
```diff
@@ -280,95 +296,122 @@
...
+                                MeasurementQuantities)
+                           [{:context-medium MeasurementContextMedium
+                             :object MeasurementObject}])
+            invalid-measurements (remove #(get measurement-index %) measurements)]
+        (seq invalid-measurements)))
+    (catch Exception e
+      (if (clojure.string/includes? (ex-message e) "Carmine connection error")
+        (do
+          (error "lookup-by-measurement found redis carmine exception. Will return nil result." e)
+          nil)
```
Comment (eereiter) [inline common-app-lib/src/cmr/common_app/services/kms_lookup.clj:416]: Remove the (do and the nil)
(error returns nil.

Please do this in the kms_lookup namespace as well.

**Review by eereiter — CHANGES_REQUESTED** [review]
Remove the (do and the nil)
(error returns nil.

**Review by eereiter — APPROVED** [review]

**Review by jceaser — APPROVED** [review]

**Pull request merged**